server/common/rest.py

```diff
@@ -215,6 +308,139 @@ def inflate(data):
     return zlib.decompress(data)
 
 
+def annotations_obs_put(request, data_adaptor):
+    if not hasattr(data_adaptor, "save_obs_annotations"):
+        return abort(HTTPStatus.NOT_IMPLEMENTED)
+
+    compressed = request.get_data()
+    if not compressed:
+        return abort(HTTPStatus.BAD_REQUEST)
+
+    try:
+        dataframe = decode_matrix_fbs(inflate(compressed))
+    except Exception as error:  # pragma: no cover - defensive logging only
+        return abort_and_log(
+            HTTPStatus.BAD_REQUEST,
+            f"Failed to decode annotations payload: {error}",
+            include_exc_info=True,
+        )
+
+    user_id = _resolve_request_user_id(request)
+
+    # Reject saves when authentication is disabled
+    if user_id is None:
+        return abort(HTTPStatus.UNAUTHORIZED)
+
+    # Reject saves when authentication is disabled
+    if user_id is None:
+        return abort(HTTPStatus.UNAUTHORIZED)
+
+    try:
+        data_adaptor.save_obs_annotations(
+            dataframe,
+            user_id=user_id,
+        )
+        return make_response(jsonify({"status": "OK"}), HTTPStatus.OK)
+    except NotImplementedError:
+        return abort(HTTPStatus.NOT_IMPLEMENTED)
+    except Exception as error:  # pragma: no cover - backend owns persistence errors
+        return abort_and_log(
+            HTTPStatus.BAD_REQUEST,
+            f"Failed to save annotations: {error}",
+            include_exc_info=True,
+        )
+
+
+def annotations_obs_category_delete(request, data_adaptor):
```

Comment:
Breadcrumb: the new DELETE/PUT handlers mirror the category/rename buttons in the UI. They expect `CELLXGENE_DISABLE_AUTH` to be off so `_resolve_request_user_id` returns something, otherwise the UI will log-and-skip the server call. Helpful to read this pair alongside the thunk code that calls them.